core/src/xtdb/query.clj

```diff
@@ -947,6 +952,22 @@
                          (mapv #(db/decode-value value-serde %) vs))]
             (bind-binding return-type value-serde tuple-idxs-in-join-order (get idx-id->idx idx-id) (not-empty values))))))))
 
+(defmethod pred-constraint 'get-valid-end-time [_ {:keys [idx-id arg-bindings tuple-idxs-in-join-order valid-time] }]
+  (let [arg-bindings (rest arg-bindings)
+        [e-var not-found] arg-bindings
+        not-found? (= 2 (count arg-bindings))
+        e-result-index (.result-index ^VarBinding e-var)]
+    (fn pred-get-valid-end-time [{:keys [value-serde index-snapshot]} idx-id->idx ^List join-keys]
+      (let [e (.get join-keys e-result-index)
+            vs (->> (db/entity-history index-snapshot (db/decode-value value-serde e) :asc {})
```

Comment:
`db/entity-history` already takes various valid-time params - do you need a separate `drop-while` here?

Reply:
Good point. I still need to drop one item in the case where the db `valid-time` coincides with the valid-time of the document. So the semantics of the `get-end-valid-time` are the next change of the entity according to the `db` except in the case where `valid-time` of the db an the document coincide.

Reply:
paired - we tend to use `(inc valid-time)` for these cases